src/realm/array_mixed.cpp

```diff
@@ -0,0 +1,352 @@
+{
+    if (value.is_null()) {
+        set_null(ndx);
+        return;
+    }
+    erase_linked_payload(ndx);
+    m_composite.set(ndx, store(value));
+}
+
+void ArrayMixed::insert(size_t ndx, Mixed value)
+{
+    if (value.is_null()) {
+        m_composite.insert(ndx, 0);
+        return;
+    }
+    m_composite.insert(ndx, store(value));
+}
+
+void ArrayMixed::set_null(size_t ndx)
+{
+    auto val = m_composite.get(ndx);
+    if (val) {
+        erase_linked_payload(ndx);
+        m_composite.set(ndx, 0);
+    }
+}
+
+Mixed ArrayMixed::get(size_t ndx) const
+{
+    int64_t val = m_composite.get(ndx);
+
+    if (val) {
+        int64_t int_val = val >> s_data_shift;
+        size_t payload_ndx = size_t(int_val);
+        DataType type = DataType((val & s_data_type_mask) - 1);
+        switch (type) {
+            case type_Int: {
+                if (val & s_payload_idx_mask) {
+                    ensure_int_array();
+                    return Mixed(m_ints.get(payload_ndx));
+                }
+                return Mixed(int_val);
+            }
+            case type_Bool:
+                return Mixed(int_val != 0);
+            case type_Float:
+                ensure_int_array();
+                return Mixed(type_punning<float>(m_ints.get(payload_ndx)));
+            case type_Double:
+                ensure_int_array();
+                return Mixed(type_punning<double>(m_ints.get(payload_ndx)));
+            case type_String: {
+                ensure_string_array();
+                REALM_ASSERT(size_t(int_val) < m_strings.size());
+                return Mixed(m_strings.get(payload_ndx));
+            }
+            case type_Binary: {
+                ensure_string_array();
+                REALM_ASSERT(size_t(int_val) < m_strings.size());
+                auto s = m_strings.get(payload_ndx);
+                return Mixed(BinaryData(s.data(), s.size()));
+            }
+            case type_Timestamp: {
+                ensure_int_pair_array();
+                payload_ndx <<= 1;
+                REALM_ASSERT(payload_ndx + 1 < m_int_pairs.size());
+                return Mixed(Timestamp(m_int_pairs.get(payload_ndx), int32_t(m_int_pairs.get(payload_ndx + 1))));
+            }
+            default:
+                break;
+        }
+    }
+
+    return {};
+}
+
+void ArrayMixed::erase(size_t ndx)
+{
+    erase_linked_payload(ndx);
+    m_composite.erase(ndx);
+}
+
+void ArrayMixed::truncate_and_destroy_children(size_t ndx)
+{
+    for (size_t i = size(); i > ndx; i--) {
+        erase_linked_payload(i - 1);
+    }
+    m_composite.truncate(ndx);
+}
+
+void ArrayMixed::move(ArrayMixed& dst, size_t ndx)
+{
+    auto sz = size();
+    size_t i = ndx;
+    while (i < sz) {
+        auto val = get(i++);
+        dst.add(val);
+    }
+    while (i > ndx) {
+        erase_linked_payload(--i);
+    }
+    m_composite.truncate(ndx);
+}
+
+size_t ArrayMixed::find_first(Mixed value, size_t begin, size_t end) const noexcept
```

Comment:
If we had the type information at the beginning of the composite value, could we use the existing index on the composite array to optimize find?

Reply:
The existing index can only be used on columns in tables.